### PR TITLE
Add preconditions for Java NotNull annotations

### PR DESCRIPTION
**Pull request by peterschrammel**
Adds support for NotNull annotations provided by some Java libraries. Unfortunately, there is no standard for that.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f1c05e825c523f1659fee515124048fe2e61c38c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115685256

**Review by kroening — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -966,6 +969,55 @@ static std::size_t get_bytecode_type_width(const typet &ty)
...
 
+code_blockt java_bytecode_convert_methodt::convert_parameter_annotations(
+  const methodt &method,
+  const java_method_typet &method_type)
+{
+  code_blockt code;
+
+  // Consider parameter annotations
+  const java_method_typet::parameterst &parameters(method_type.parameters());
+  const bool skip_this = !parameters.empty() && parameters[0].get_this();
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: method_type.has_this()

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b11ffc8324e77ee971853fe1ec4411963dd381e2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115719919

File: jbmc/regression/jbmc/parameter-annotation-not-null/test_array.desc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+CORE
+Main.class
+--function Main.test2 --classpath `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+\[not-null-annotation-check\.1\] line \d+ Not null annotation check: FAILURE
+\[not-null-annotation-check\.2\] line \d+ Not null annotation check: SUCCESS
```
Comment (thk123) [inline jbmc/regression/jbmc/parameter-annotation-not-null/test_array.desc:8]: :question: Why so few assert checks? Surely for this there should be 3 (for x, z, s) of which x and z can both fail, but y must succeed?

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 3]
```diff
@@ -966,6 +969,54 @@ static std::size_t get_bytecode_type_width(const typet &ty)
...
+        parameters[param_index].get_identifier();
+      const symbolt &param_symbol = symbol_table.lookup_ref(param_identifier);
+      const auto param_type =
+        type_try_dynamic_cast<pointer_typet>(param_symbol.type);
+      if(param_type)
+      {
+        code_assertt code_assert(notequal_exprt(
+          param_symbol.symbol_expr(), null_pointer_exprt(*param_type)));
+        source_locationt check_loc = method.source_location;
+        check_loc.set_comment("Not null annotation check");
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:1007]: :pick: Include the name of the parameter in the name of the check?

**Review by thk123 — COMMENTED** [review]
Was interested so had a quick look. Does this have TG implication (e.g. do these get converted to assumes when this f.u.t. has these annotations?

**Pull request merged**